helper/schema/resource_diff.go

```diff
@@ -234,6 +236,19 @@ func (d *ResourceDiff) clear(key string) error {
 	return nil
 }
 
+// GetDiffAttributes helps to implement resourceDiffer
+// where we need to act on all nested fields
+// without calling out each field separately
+func (d *ResourceDiff) GetDiffAttributes(prefix string) map[string]*terraform.ResourceAttrDiff {
```

Comment:
> hard-code the keys you want to mark as `ForceNew`

The schema is really deeply nested and long and it is likely to grow in the future:
https://github.com/terraform-providers/terraform-provider-kubernetes/blob/master/kubernetes/schema_volume_source.go#L14
so checking each field seems like a rather tedious approach.

> maybe in the example that you are giving, you could probably just either walk the schema itself directly

That means we'd have to take the schema and try to "guess" how does it (or may) translate into keys (indexes) in the diff, which is difficult to do specifically for TypeList/TypeSet/TypeMap. It also seems like that's a logic that should be somewhere in core anyway - the provider IMO shouldn't have to traverse its own schema and know how to translate it to "diff-like format" in order to call `d.Get(key)` or `d.GetChange(key)`.

However I understand what you're saying and I did have similar feelings when cobbling the PR. It does not feel right to expose the diff like this, so thanks for confirming my thoughts. 😃

I'm playing with two ideas at this point:

 - `func (d *ResourceDiff) ForceNewPrefix(prefix) error` which would basically go through the schema and mark all nested fields (those which have the given prefix in state) as `ForceNew`
 - `func (d *ResourceDiff) GetChangedKeys(prefix) []string` - here we'd be still exposing the diff, but only keys, so the bare minimum we need to (in order to call `d.Get()` and `d.ForceNew()` on them), not the whole `map[string]*terraform.ResourceAttrDiff`

What do you think?

Reply:
Just wanted to chime in here and agree with @vancluever's reservation... we should not expose _any_ `terraform` package types in the `helper/schema` API because our future grpc-protocol will not be able to marshal these.

(There is one existing violation of that rule in the schema migration mechanism which we'll be designing a special workaround for, but we should avoid introducing any new dependencies like this here.)

Reply:
@radeksimko I like both of those ideas, and almost think `GetChangedKeys` would be nice for `ResourceData` to have too if it weren't for the fact that the diff reader is not available in `Read`.

My one concern for `GetChangedKeys` is feature overlap with `HasChange`, but that's kind of a gut reaction over a logical one, especially considering you could debate that `GetChange` has overlap as well by the same logic. As well, I'm wondering if two versions of `GetChangedKeys` would be nice where you didn't have to supply the prefix and you could get the _entire_ set of changes resource-wide, versus having to work with a possibly semantically awkward `GetChangedKeys("")`.

But yeah, I think that's a great idea, and barring @apparentlymart saying something otherwise I don't think that exposing just the keys from the diff is risky in regards to any future changes in the API.